conjure-undertow-lib/src/main/java/com/palantir/conjure/java/undertow/lib/EndpointWrapper.java

```diff
@@ -0,0 +1,22 @@
+/*
+ * (c) Copyright 2019 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.conjure.java.undertow.lib;
+
+public interface EndpointWrapper {
+
+    Endpoint wrap(Endpoint endpoint);
```

Comment:
What’s the use-case for being able to totally transform Endpoint definitions?

Reply:
In most case you are only interested to wrap the inner handler but in this case it would still be pretty convenient:`endpoint -> endpoint.mapHandler(handler -> { ...})`
1. It makes the api more intuitive/consistent
2. it gives us flexibility for the future if we add new fields to Endpoint
3. There are some actual case where you might be interested to rewrite the path and maybe the name (for instance if you want to have a v2 of an api that change only a few things, add "2" to the ServiceName and change the path.

Reply:
My preference is to avoid utility functionality in conjure-undertow-lib, which should be the simplest possible target for our generated code. Similarly I don't think we want to make it convenient to wrap endpoints. It's a necessity for some scenarios, but also something to be avoided unless there are no other options.

The problem we need to solve is providing a replacement for servlet and jersey filter mechanisms, for example providing a filter which can respond with 503 until the service reports that it has been initialized.

Something along these lines will be more specific:
```java
interface EndpointFilter {

    /** 
     * Returns a handler wrapping {@link Endpoint#handler()} if the given endpoint is matched,
     * otherwise an empty optional if this filter does not apply to the endpoint.
     */
    Optional<HttpHandler> filter(Endpoint endpoint);
}
```

We may want to implement that in a new module in conjure-java-runtime-api so we can avoid exposing an api dependency on conjure-undertow-runtime

Reply:
I have mixed feelings about this (not the part about the package location but EndpointFilter).

It seems weird to me that we call it a `Filter` which is a jersey concept when this is just a plain HandlerWrapper. It also requires consumers to infer that they are expected to wrap the Endpoint#handler() but actually nothing forbid them to just return a completely non-related `HttpHandler`

Reply:
It maps closely to a jersey DynamicFeature, which is one of the most powerful concepts from jersey. It makes it easy for implementors to pre-compute more than a standard servlet filter, which is what I'm going for. I think it's a bit dangerous if we return `Endpoint` because it allows implementors to remove the deprecated field or modify the http methd. In the future if we add new functionality to Endpoint, we may need to expand the filter/wrapper concept, but I think that is likely to be safer than allowing full modification of the Endpoint.

Agreed that there's nothing to forbid returning a completely different handler, but that's no different from returning an Endpoint. There's no way around this if we want to allow try/finally blocks around the implementation, which allows us to code more safely.

Reply:
I still don't see the benefits of mapping to a jersey concept when we have the benefits in conjure-undertow of being able to represent Endpoint and UndertowService as nice static data containers that can be seen as Functors and for whose the map function is a natural extension.

The so-called filters on a Service is easier to understand as a wrapper to all endpoints than as "the equivalent of the jersey concept of applying pre and post subroutine to jax-rs resources".

The most natural signature of an Endpoint transformation is `Endpoint -> Endpoint`. It is very clear what it is doing and it seems unlikely that users would be confused by its use. If they really want to change the http method, it is weird, but it is also ultimately it is their choice. There is no way they would change the http method without realising it.

That being said, if it really bother you, could we consider transforming `EndpointWrapper` to `EndpointHandlerWrapper { HttpHandler wrap(Endpoint) }` and UndertowService to 

```
UndertowService {
   ...
  UndertowService mapHandler(EndpointHandlerWrapper) 
}
```